sos/plugins/__init__.py

```diff
@@ -738,10 +738,13 @@ def _copy_symlink(self, srcpath):
 
     def _copy_dir(self, srcpath):
         try:
-            for afile in os.listdir(srcpath):
+            for path in os.listdir(srcpath):
                 self._log_debug("recursively adding '%s' from '%s'"
-                                % (afile, srcpath))
-                self._do_copy_path(os.path.join(srcpath, afile), dest=None)
+                                % (path, srcpath))
+                if os.path.isdir(path):
+                    self._copy_dir(srcpath)
```

Comment:
This test and branch sounds wrong to me:

- `path` is always an item inside `srcpath` dir, so it rarely would be a directory itself. So this branch would (almost) never be executed. 
- even if the branch would be executed, we would recursively call the same method with the same argument

So one approach is to have here:

    if os.path.isdir(os.path.join(srcpath, path)):
        self._copy_dir(os.path.join(srcpath, path))
    else:
        self._do_copy_path(os.path.join(srcpath, path))

(optionally using a variable for the `join`). But we should simply call *just*

    self._do_copy_path(os.path.join(srcpath, path))

every time (as the code executes even now practically every time). Otherwise, if we would recursively call `_copy_dir` directly, I am afraid we would: 1) skip `_is_forbidden_path` for the subdir (try collecting `/foo` dir but forbid `/foo/bar` subdir), 2) skip adding `/foo/bar` to the archive.

So I suggest removing the 3 lines of practically dead code.

Reply:
Hm, my suggestion reverts back @bmr-cymru's whole change in this method, that makes me less confident.

Reply:
That was my mistake; the important part of the change is correcting the argument passed to `_copy_dir()` from `_do_copy_path()`, not the handling of sub-directories in the method itself.